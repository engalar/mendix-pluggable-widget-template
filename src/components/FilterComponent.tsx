import { CSSProperties, ReactElement, useCallback, useEffect, useRef, useState } from "react";
import { FilterSelector } from "../piw-utils-internal/widgets";
import { debounce } from "../piw-utils-internal/functions";

import { DefaultFilterEnum } from "../../typings/DatagridTextFilterProps";
import classNames from "classnames";

interface FilterComponentProps {
    adjustable: boolean;
    className?: string;
    defaultFilter: DefaultFilterEnum;
    delay: number;
    id?: string;
    placeholder?: string;
    tabIndex?: number;
    screenReaderButtonCaption?: string;
    screenReaderInputCaption?: string;
    styles?: CSSProperties;
    updateFilters?: (value: string, type: DefaultFilterEnum) => void;
    value?: string;
}

export function FilterComponent(props: FilterComponentProps): ReactElement {
    const [type, setType] = useState<DefaultFilterEnum>(props.defaultFilter);
    const [value, setValue] = useState("");
    const [valueInput, setValueInput] = useState("");
    const inputRef = useRef<HTMLInputElement | null>(null);

    useEffect(() => {
        if (props.value) {
            setValueInput(props.value);
            setValue(props.value);
        }
    }, [props.value]);

    useEffect(() => {
        props.updateFilters?.(value, type);
    }, [value, type]);

    const onChange = useCallback(
        debounce((value: string) => setValue(value), props.delay),
        [props.delay]
    );

    const focusInput = useCallback(() => {
        if (inputRef.current) {
            inputRef.current.focus();
        }
    }, [inputRef]);

    return (
        <div
            className={classNames("filter-container", props.className)}
            data-focusindex={props.tabIndex ?? 0}
            style={props.styles}
        >
            {props.adjustable && (
                <FilterSelector
                    ariaLabel={props.screenReaderButtonCaption}
                    // id={props.id}
                    defaultFilter={props.defaultFilter}
                    onChange={useCallback(
                        type => {
                            setType(prev => {
                                if (prev === type) {
                                    return prev;
                                }
                                focusInput();
                                return type;
                            });
                        },
                        [focusInput]
                    )}
                    options={
                        [
                            { value: "contains", label: "Contains" },
                            { value: "startsWith", label: "Starts with" },
                            { value: "endsWith", label: "Ends with" },
                            { value: "greater", label: "Greater than" },
                            { value: "greaterEqual", label: "Greater than or equal" },
                            { value: "equal", label: "Equal" },
                            { value: "notEqual", label: "Not equal" },
                            { value: "smaller", label: "Smaller than" },
                            { value: "smallerEqual", label: "Smaller than or equal" },
                            { value: "empty", label: "Empty" },
                            { value: "notEmpty", label: "Not empty" }
                        ] as Array<{ value: DefaultFilterEnum; label: string }>
                    }
                />
            )}
            <input
                aria-label={props.screenReaderInputCaption}
                className={classNames("form-control", { "filter-input": props.adjustable })}
                disabled={type === "empty" || type === "notEmpty"}
                onChange={e => {
                    setValueInput(e.target.value);
                    onChange(e.target.value);
                }}
                placeholder={props.placeholder}
                ref={inputRef}
                type="text"
                value={valueInput}
            />
        </div>
    );
}
