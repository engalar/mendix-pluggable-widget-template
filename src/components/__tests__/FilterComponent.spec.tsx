import { render, shallow } from "enzyme";
import { createElement } from "react";
import { FilterComponent } from "../FilterComponent";

jest.useFakeTimers();

describe("Filter component", () => {
    it("renders correctly", () => {
        const component = render(<FilterComponent adjustable defaultFilter="contains" delay={500} />);

        expect(component).toMatchSnapshot();
    });

    it("renders correctly when not adjustable by user", () => {
        const component = render(<FilterComponent adjustable={false} defaultFilter="contains" delay={500} />);

        expect(component).toMatchSnapshot();
    });

    it("renders correctly with aria labels", () => {
        const component = render(
            <FilterComponent
                adjustable
                screenReaderButtonCaption="my label"
                screenReaderInputCaption="my label"
                defaultFilter="contains"
                delay={500}
            />
        );

        expect(component).toMatchSnapshot();
    });

    it("calls updateFilters when value changes", () => {
        const updateFiltersHandler = jest.fn();
        const component = shallow(
            <FilterComponent adjustable defaultFilter="contains" delay={500} updateFilters={updateFiltersHandler} />
        );

        const input = component.find("input");
        input.simulate("change", { target: { value: "test" } });

        expect(updateFiltersHandler).toBeCalled();
    });

    it("debounces calls for updateFilters when value changes", () => {
        const updateFiltersHandler = jest.fn();
        const component = shallow(
            <FilterComponent adjustable defaultFilter="contains" delay={500} updateFilters={updateFiltersHandler} />
        );

        // Initial call with default filter
        expect(updateFiltersHandler).toBeCalledTimes(1);

        const input = component.find("input");
        input.simulate("change", { target: { value: "test" } });
        jest.advanceTimersByTime(499);
        input.simulate("change", { target: { value: "test2" } });
        input.simulate("change", { target: { value: "test3" } });
        jest.advanceTimersByTime(500);

        expect(updateFiltersHandler).toBeCalledTimes(2);

        input.simulate("change", { target: { value: "test" } });
        jest.advanceTimersByTime(500);

        expect(updateFiltersHandler).toBeCalledTimes(3);
    });
});
