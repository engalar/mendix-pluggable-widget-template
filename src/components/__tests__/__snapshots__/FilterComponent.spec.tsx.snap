// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter component renders correctly 1`] = `
<div
  class="filter-container"
  data-focusindex="0"
>
  <div
    class="filter-selector"
  >
    <div
      class="filter-selector-content"
    >
      <button
        aria-controls="undefined-filter-selectors"
        aria-expanded="false"
        aria-haspopup="true"
        class="btn btn-default filter-selector-button button-icon contains"
      >
         
      </button>
    </div>
  </div>
  <input
    class="form-control filter-input"
    type="text"
    value=""
  />
</div>
`;

exports[`Filter component renders correctly when not adjustable by user 1`] = `
<div
  class="filter-container"
  data-focusindex="0"
>
  <input
    class="form-control"
    type="text"
    value=""
  />
</div>
`;

exports[`Filter component renders correctly with aria labels 1`] = `
<div
  class="filter-container"
  data-focusindex="0"
>
  <div
    class="filter-selector"
  >
    <div
      class="filter-selector-content"
    >
      <button
        aria-controls="undefined-filter-selectors"
        aria-expanded="false"
        aria-haspopup="true"
        aria-label="my label"
        class="btn btn-default filter-selector-button button-icon contains"
      >
         
      </button>
    </div>
  </div>
  <input
    aria-label="my label"
    class="form-control filter-input"
    type="text"
    value=""
  />
</div>
`;
